pkg/api/unversioned.go

```diff
@@ -22,16 +22,49 @@ import (
 
 // This file contains API types that are unversioned.
 
-// APIVersions lists the api versions that are available, to allow
-// version negotiation. APIVersions isn't just an unnamed array of
-// strings in order to allow for future evolution, though unversioned
+// APIVersions lists the versions that are available, to allow
+// clients to discover the API. This is returned at /api
 type APIVersions struct {
+	// Versions are the api versions that are available.
 	Versions []string `json:"versions"`
 }
 
+type APIGroupList struct {
+	Groups []APIGroup `json:"groups"`
+}
+
+type APIGroup struct {
+	GroupName        string    `json:"groupName"`
+	Versions         []Version `json:"versions"`
```

Comment:
For Openshift I'd like to embed uris here as soon as we sort out component
registration so we can do relative or absolute pointers from versions to
target (and same for group)

On Sep 16, 2015, at 8:21 PM, Chao Xu <notifications@github.com> wrote:

In pkg/api/unversioned.go
<https://github.com/kubernetes/kubernetes/pull/13955#discussion_r39700847>:

>  	Versions []string `json:"versions"`
>  }
>
> +type APIGroupList struct {
> +	Groups []APIGroup `json:"groups"`
> +}
> +
> +type APIGroup struct {
> +	GroupName        string    `json:"groupName"`
> +	Versions         []Version `json:"versions"`

Quoting @smarterclayton <https://github.com/smarterclayton>: "I expect we
will add url or uri to both version and
group in the future"

So we should use Version struct to keep the extensibility.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13955/files#r39700847>.